### ISSUE TITLE
CSV Upload

### ISSUE THREAD

**Issue by vanwars** [LocalGround/localground#79]
Please make a python script that handles inserting CSV data into the database:

### Step 1. The user provides the script with:

1. The CSV file
2. The destination table id (where they want their data to be inserted to). 
3. A project id
4. An owner id
5. Regarding: column mappings, assume, for now that the column headers have the exact same as the "col_name" of the table.

Question: Should we make a queue table to hold these CSVs, like another API Endpoint? I kinda think we should, but whatever you think.

### Step 2. Process data
1. Iterate through CSV rows, convert each text cell to typed variable that DB is expecting, and insert. There may already be a Django or Python library that does this.
2. If there are errors, flag the rows that have caused errors. Perhaps in the original file, you could just make a new column w/the error message?
3. Send email indicating success of failure. If failure, return the error message w/the original CSV + error messages.

### Tips
Tables are our most complicated model, because they are user-generated, and therefore every table as a different schema. So, to instantiate a new Table, you need to do the following:
```
from localground.apps.site import models
form = models.Form.objects.get(id=form_id)
new_record = form.TableModel()
new_record.project = the_project
new_record.owner = the_owner
...
new_record.save()
```
An example can be found in /localground/apps/site/tests/views/forms.py

**Comment by awesomescot** [comment]
So for now should step 1 be an api endpoint?  That is where the CSV could be uploaded, then that kicks off the script to process it.

**Comment by vanwars** [comment]
Yes, I think Step 1 should be an endpoint that stores all the required inputs, and saves the CSV to disk.

**Comment by awesomescot** [comment]
Ok, I'm thinking about how to implement this.  It would seem that we don't really need a model and db table for this, I think this is what you're getting at with your question above?  So we could just have an api endpoint that gets the CSV and other parameters, saves the CSV to disk, and calls the async method in celery.  Celery will keep track of how the task is doing in it's result backend.  I guess there are a few questions here for me, firstly can you forsee a situation where we would want our own table that keeps all  csv data?  Like will there be a frontend to view all csv?  I imagine that we could use the result backend for that also, but it would probably be more difficult for us right now.  Then there is the question of how do I create an api endpoint without a model?  Anyway, I do think that using the result backend is probably the more correct way, but it might be a lot easier to have our own model, because then I can basically just copy what's happening in the barcoded serializer.  So I might start with that.  Feel free to give me your input though.

**Comment by vanwars** [comment]
Thanks for these questions. Very helpful for thinking this through. Responses below:

> an you forsee a situation where we would want our own table that keeps all csv data? Like will there be a frontend to view all csv?

Hmm, I think we probably should...as a user you'd probably want to know whether your file was successfully updated and its' current status if it's going to trigger an asynchronous process.
> Anyway, I do think that using the result backend is probably the more correct way, but it might be a lot easier to have our own model, because then I can basically just copy what's happening in the barcoded serializer. So I might start with that. 

That is exactly what I would do, but as you said, it's because I've never used the result backend. So I guess if it's possible for us to store custom parameters in the result backend and allow the user to view it, then let's use the result backend. Otherwise, let's make another model. And, since we may eventually provide support for import for other formats (like Shapefiles), maybe call the model something generic.

**Comment by awesomescot** [comment]
@vanwars how generic should this be?  Currently it's class CSVFile.  I can make it more generic, I just need to know how much.  Are the other files still CSVs or are they a different format all together?

**Comment by awesomescot** [comment]
Ok, so I'm running into issues with the instance view.  Firstly should you only be able to see csv files that you own?  Secondly I'm a little confused how the id from the url regex gets passed to the view, r'^csv/(?P<pk>[0-9]+)/$'.  I guess it gets the queryset and then filters by id?  Seems like there is some magic happening here.

**Comment by vanwars** [comment]
> Firstly should you only be able to see csv files that you own? 

In Django, there is an object called a "Manager" that allows you to create custom queryset methods. See /localground/apps/site/managers/ directory for details.


We're currently changing the permissions model for this, but for now, just
Secondly I'm a little confused how the id from the url regex gets passed to the view, r'^csv/(?P[0-9]+)/$'. I guess it gets the queryset and then filters by id? Seems like there is some magic happening here.
